staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,617 @@
+		if !cached.stale {
+			cached = cachedResult{
+				discovery:   cached.discovery,
+				etag:        cached.etag,
+				lastUpdated: now,
+				stale:       true,
+			}
+			dm.cachedResults[info.service] = cached
+		}
+
+		klog.Infof("DiscoveryManager: Failed to download discovery for %v: %v %s",
+			info.service.String(), writer.respCode, writer.data)
+		return cached, fmt.Errorf("service %s returned non-success response code: %v",
+			info.service.String(), writer.respCode)
+	}
+}
+
+// Try to sync a single APIService.
+func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
+	dm.servicesLock.RLock()
+	info, exists := dm.apiServices[apiServiceName]
+	dm.servicesLock.RUnlock()
+
+	gv := helper.APIServiceNameToGroupVersion(apiServiceName)
+	mgv := metav1.GroupVersion{Group: gv.Group, Version: gv.Version}
+
+	if !exists {
+		// apiservice was removed. remove it from merged discovery
+		dm.mergedDiscoveryHandler.RemoveGroupVersion(mgv)
+		return nil
+	}
+
+	// Lookup last cached result for this apiservice's service.
+	now := time.Now()
+	cached, err := dm.fetchFreshDiscoveryForService(mgv, info)
+
+	dm.servicesLock.Lock()
+	info.lastReconciled = now
+	dm.apiServices[apiServiceName] = info
+	dm.servicesLock.Unlock()
+
+	if err != nil {
+		// There was an error fetching discovery for this APIService.
+		// Just use empty GV to mark that GV exists, but no resources.
+		//
+		// TODO: Maybe also stick in a status for the version the error?
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
+			Version: gv.Version,
+		})
+		return err
+	}
+
+	// Check cache entry for this groupversion and insert it into the
+	// document if present
+	if entry, exists := cached.discovery[mgv]; exists {
+		//!TODO: mark with staleness of cache
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+	} else {
+		// Use empty GV, since there is an APIService for it
+		//!TODO: mark with staleness of cache
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
+			Version: gv.Version,
+		})
+	}
+
+	return nil
+}
+
+// Spwans a goroutune which waits for added/updated apiservices and updates
+// the discovery document accordingly
+func (dm *discoveryManager) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting ResourceDiscoveryManager")
+
+	// Spawn workers
+	// These workers wait for APIServices to be marked dirty.
+	// Worker ensures the cached discovery document hosted by the ServiceReference of
+	// the APIService is at least as fresh as the APIService, then includes the
+	// APIService's groupversion into the merged document
+	for i := 0; i < 2; i++ {
+		go func() {
+			for {
+				next, shutdown := dm.dirtyAPIServiceQueue.Get()
+				if shutdown {
+					return
+				}
+
+				if err := dm.syncAPIService(next.(string)); err != nil {
+					dm.dirtyAPIServiceQueue.AddRateLimited(next)
+				} else {
+					dm.dirtyAPIServiceQueue.AddAfter(next, 5*time.Second)
```

Comment:
it has been a while, but as I remember the workqueue code, we don't de-dupe the addafter, so by doing this I would expect the queue to get busier and busier over time.  Are you trying to resync every service every 5s to check to see if discovery changed?  If so, I suggest using a separate resync loop to queue and use the "normal" controller logic here.

Reply:
ack, received the same comment from a couple of internal reviewers and am already working on fixing this.